### PR TITLE
Load X11 dynamically

### PR DESCRIPTION
**Pull request by deralmas**
The loaders have been generated through @hpvb's dynload-wrapper, although they had to be heavily handpatched to workaround some already reported issues with it. I added a note to each generated file to account for that.

As GLAD uses X11 stuff directly, I had to define the `GLAD_GLX_NO_X11` macro to not let do it that, and handle myself the display loading and screen handling part myself, which wasn't that hard but it's still something worth saying.

I plan to improve greatly the X11 backend (including this aspect) but, as the release isn't that far and I'm also working on the Wayland backend, this will do for now, I hope.

As the patches are pretty big, I made a zip with the modified .so wrappers and the patches used to make them: [x11-dynwrappers.zip](https://github.com/godotengine/godot/files/10147096/x11-dynwrappers.zip)

### CODE REVIEW

File: platform/linuxbsd/x11/xext-so_wrap.c [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+// generated by ./generate-wrapper.py 0.3 on 2022-11-30 13:06:13
+// flags: ./generate-wrapper.py --include /usr/include/X11/extensions/Xext.h --sys-include <X11/extensions/Xext.h> --include /usr/include/X11/extensions/shape.h --sys-include <X11/extensions/shape.h> --soname libXext.so.6 --init-name xext --output-header xext-so_wrap.h --output-implementation xext-so_wrap.c
+// NOTE: This has been heavily handpatched to workaround some generator issues
+// that have already been reported and are being worked on. Namely, some complex
+// pointer types have been properly carried over, some missing includes were
+// added and lots of duplicate symbols have been removed, although perhaps not
+// all of them.
+#include <stdint.h>
+
+// HANDPATCH: Needed for successfull compilation.
```
Comment (akien-mga) [inline platform/linuxbsd/x11/xext-so_wrap.c:12]: "successful"

Comment (akien-mga) [reply platform/linuxbsd/x11/xext-so_wrap.c:12]: I've also been struggling with this one for years. So confusing :D

**Comment by akien-mga** [comment]
It might fail on the older Ubuntu from CI if the wrappers were generated from a more recent X11.
For maximum compatibility we likely need to generate the wrappers against an older X11 so that they don't include any of the recent APIs (since we still need these symbols to be in the actual shared library on the user's system).

**Comment by akien-mga** [comment]
> the thing is that I heavily handpatched this stuff and redoing all that work again would be quite a bit of a chore. 

You should probably generate a `.patch` file with the differences between the output from the generator and what you had to do to make it work. So we can redo those changes whenever we want to update - and possibly to be able to reapply those changes on an older version of X11 if this is confirmed to be needed.

You can indeed also try to remove APIs but you'd have to check the diff between headers down to the oldest version we want to support to see what was added and should be removed.

**Comment by dsnopek** [comment]
> You should probably generate a .patch file with the differences between the output from the generator and what you had to do to make it work.
>> I already described most of the important changes done in the note and I've already reported most of those issues to hpvb. Perhaps a patch would be useful for improving the generator, but I doubt that it would be that useful (as it's mostly removing duplicate symbols, there are actually very few parsing issues) and I don't think that it'd be worth to keep in the source tree.

A patch file in the source tree would be most useful when someone else (ie. not you) wants to update the code with the latest output from the generator. The notes are nice, but being able to see the actual diff is clearer than any description. (It also makes this PR potentially easier to review, because the reviewers can see exactly what changes you had to make.)

Also, if the issues were reported hpvb public issues, including links to those in a text file or comments in the source code would also be helpful. That way, if someone is updating the code they can check the issues as a away to see if the changes you made are still necessary.

**Comment by akien-mga** [comment]
Hm OK the patches are huge, I didn't expect so many changes. This should really be fixed in the wrapper tool as this will be hard to maintain going forward.

But I guess we can remove the patches from this PR then, and maybe just upload them to the PR description as reference.

The separate folder for wrappers is still a nice change.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by omar-polo** [comment]
This seems to have broken the build on OpenBSD.

The various `*-so_wrap.c` files includes the major in and this can't work.  In -CURRENT libX11 is, for example, `libX11.so.18.0`, but this is (of course) subject to change on the next X11 update (if they break the ABI somehow)

I guess that these needs to be regen'd without the major?

**Comment by akien-mga** [comment]
The major is needed, `.so` files are symlimks only present when development packages are installed.

**Comment by omar-polo** [comment]
> Eventually, for platforms like *BSD and Nix which do not or cannot handle portable binaries, we should re-add build options to skip the wrappers and link the system libraries directly (like we have with the builtin_* options for non-dlopen'ed libraries).

If this "portable binaries" means setting ABI in stone I fear this won’t work at all on OpenBSD.  Source-code compatibility is what is generally guaranteed, not binary.  (we had at least two major bump of libc and libX11 in the last two years, and many more minor bumps.)

Would love to have an option to link to libraries as Godot 3 did, but I fear I'm not skilled enough to work on such a patch, apologies :(

**Comment by akien-mga** [comment]
> If Godot properly linked to libX11, or used dlopen() without the major, then the shipped package will be automatically updated when libX11 changes, without having to patch the sources. From the POV of a package maintainer this would be painless and reduce friction.

Note that if Godot linked to libX11, you'd still need to recompile the Godot package whenever the libX11 soname changes in OpenBSD. So if a recompile is needed either way, currently you could patch the file to hardcode OpenBSD's soname (and hope it works against the parts of the API Godot relies on). It's of course not an elegant solution, just noting it as something that could work for now.

**Comment by omar-polo** [comment]
That's true, but bulk builds run every day so it's not a problem.  Libraries gets updated all the time and packages rebuit/linked to the new versions.  if a hypothetical new X11 update introduces a breaking change i'll wake up one morning finding a (failed) build log in my inbox :)
On the other hand, this yolo-dlopen-at-runtime approach means I need to periodically rebuild and test Godot.  FWIW i prefer breaking at build-time if possible.

Now, this is just my preference from the POV of a package maintainer, I'm sure that if you've introduced this mechanisms there were sound reasons for it and I'm not arguing that.  (to be fair i've never written software that supports the wide range of OSes that Godot does!)

I've seen the PR to make optional the use of these wrappers and I'm about to test it.  Thanks a lot for considering this use case too and working for making it possible! :)

**Pull request merged**